### ISSUE TITLE
AXI DMA: Abysmal Performance

### ISSUE THREAD

**Issue by josephzavodny** [Xilinx/linux-xlnx#62]
About a year ago I wrote a high performance driver for streaming data between the PS and PL using AXI DMA.  My driver queues scatter gather transactions with 4K blocks(entire usespace pages).  On the 3.14 kernel I was getting about 250 MB/s throughput.

Last night I decided to pull your master branch and move to the 4.0 kernel.  After updating my driver to use the new interface for acquiring and configuring DMA channels, I reran my benchmark.  With the 4.0 kernel I am only getting 0.69 MB/s.  That is less than 700 kilobytes/second.  On top of that, both Arm cores have 100% cpu usage, with all that time being spent running kernel code.

**Comment by michalsimek** [comment]
can you please send me your testing code?

**Comment by josephzavodny** [comment]
Unfortunately I don't think that will be possible.  I am working to narrow down the problem and intend to submit a patch once I have a fix.

**Comment by ismell** [comment]
Just asking the obvious question. Do you have any logging turned on? Does dmesg show anything strange?

**Comment by josephzavodny** [comment]
I don't see any error messages in dmesg.  I built axidmatest as a module and it seems to take longer to load/run than it used to.  The test succeeds the first time it is loaded but has errors if it is loaded more than once.

I have also noticed that I am getting a lot fewer interrupts than I used to- it looks like the driver is ignoring the DMA_PREP_INTERRUPT flag that I am passing to device_prep_slave_sg().

**Comment by josephzavodny** [comment]
I tracked the source of the problem to xilinx_dma_start().  This function sets the runstop bit in the control register and then waits for the halted bit in the status register to go low by polling the register.

At least that is what is supposed to happen.  A pair of parentheses were removed by mistake in commit 7a16b00c65a4862b6d9c3191426ac8297d20acc4.  This error causes the condition to only evaluate true if the hardware has not started.  If the hardware has started, it will poll the register 1 million times waiting for a condition that will never become true.

To make matters worse this is done while holding a spinlock, so as soon as the DMA transfer completes and the hardware triggers an interrupt, the interrupt handler is going to try to take the lock and will be forced to spin until the the other CPU is done polling the status register.  Which is going to take quite a while considering that involves performing 1 million reads on the AXI bus.

We should probably make that number much smaller considering that the dma hardware usually starts before the first iteration of the loop.  Not to mention the fact that the code holds a spinlock and has the potential to lock up the system for extended periods of time if the dma hardware fails to start.

I will be submitting a patch shortly.  With the fix my benchmark performance has improved from 0.69 MB/s to 290MB/s which is considerably better than what I was getting on kernel 3.14.  That said I am still experiencing crashes, system lockups, and corrupted DMA buffers so the driver is far from stable.
